typescript/mcap/src/v0/Mcap0IndexedReader.ts

```diff
@@ -1,11 +1,216 @@
+    }
+    const cursor = this.messageIndexCursors.peek();
+    if (!cursor) {
+      throw new Error(
+        `Unexpected popMessage() call when no more messages are available, in chunk at offset ${this.chunkIndex.chunkStartOffset}`,
+      );
+    }
+    const record = cursor.records[cursor.index]!;
+    const [logTime] = record;
+    if (this.startTime != undefined && logTime < this.startTime) {
+      throw new Error(
+        `Encountered message with logTime (${logTime}) prior to startTime (${this.startTime}) in chunk at offset ${this.chunkIndex.chunkStartOffset}`,
+      );
+    }
+    if (
+      cursor.index + 1 < cursor.records.length &&
+      (this.endTime == undefined || cursor.records[cursor.index + 1]![0] <= this.endTime)
+    ) {
+      cursor.index++;
+      this.messageIndexCursors.replace(cursor);
+    } else {
+      this.messageIndexCursors.pop();
+    }
+    return record;
+  }
+
+  /** FIXME document */
+  async loadMessageIndexCursors(readable: IReadable): Promise<void> {
+    this.messageIndexCursors = new Heap((a, b) => {
+      const logTimeA = a.records[a.index]?.[0];
+      const logTimeB = b.records[b.index]?.[0];
+      if (logTimeA == undefined) {
+        return 1;
+      } else if (logTimeB == undefined) {
+        return -1;
+      }
+      return Number(logTimeA - logTimeB);
+    });
+
+    let messageIndexStartOffset: bigint | undefined;
+    let relevantMessageIndexStartOffset: bigint | undefined;
+    for (const [channelId, offset] of this.chunkIndex.messageIndexOffsets) {
+      if (messageIndexStartOffset == undefined || offset < messageIndexStartOffset) {
+        messageIndexStartOffset = offset;
+      }
+      if (!this.relevantChannels || this.relevantChannels.has(channelId)) {
+        if (
+          relevantMessageIndexStartOffset == undefined ||
+          offset < relevantMessageIndexStartOffset
+        ) {
+          relevantMessageIndexStartOffset = offset;
+        }
+      }
+    }
+    if (messageIndexStartOffset == undefined || relevantMessageIndexStartOffset == undefined) {
+      return;
+    }
+
+    // Future optimization: read only message indexes for given channelIds, not all message indexes for the chunk
+    const messageIndexEndOffset = messageIndexStartOffset + this.chunkIndex.messageIndexLength;
+    const messageIndexes = await readable.read(
+      relevantMessageIndexStartOffset,
+      messageIndexEndOffset - relevantMessageIndexStartOffset,
+    );
+    const messageIndexesView = new DataView(
+      messageIndexes.buffer,
+      messageIndexes.byteOffset,
+      messageIndexes.byteLength,
+    );
+
+    let offset = 0;
+    for (
+      let result;
+      (result = parseRecord({ view: messageIndexesView, startOffset: offset, validateCrcs: true })),
+        result.record;
+      offset += result.usedBytes
+    ) {
+      if (result.record.type !== "MessageIndex") {
+        continue;
+      }
+      if (
```

Comment:
Maybe less nesting if you flip this condition around and continue in it rather than put the main logic in it?